tsdb/index/inmem/inmem.go

```diff
@@ -189,13 +189,23 @@ func (i *Index) MeasurementIterator() (tsdb.MeasurementIterator, error) {
 func (i *Index) CreateSeriesListIfNotExists(seriesIDSet *tsdb.SeriesIDSet, measurements map[string]int,
 	keys, names [][]byte, tagsSlice []models.Tags, opt *tsdb.EngineOptions, ignoreLimits bool) error {
 
+	// Verify that the series will not exceed limit.
+	if !ignoreLimits {
```

Comment:
since we're returning from here and we have to track multiple unlocks, maybe this is better as a separate function or a function literal so that we can use defer to ensure we cleanup like:

```

	if !ignoreLimits {
		i.mu.RLock()
		if max := opt.Config.MaxSeriesPerDatabase; max > 0 && len(i.series)+len(keys) > max {
			i.mu.RUnlock()
			return errMaxSeriesPerDatabaseExceeded{limit: opt.Config.MaxSeriesPerDatabase}
		}
		i.mu.RUnlock()
	}
```

Reply:
I initially had this wrapped in a func with a `defer` but it added more lines of code. This is also performance sensitive code executes with every write and `defer` has a cost. We'd also need to make this change on `1.8` branch first and then backport.